auth/postgres/groups.go

```diff
@@ -90,134 +82,178 @@ func (gr groupRepository) Save(ctx context.Context, g groups.Group) (groups.Grou
 	dbu := dbGroup{
 		ID: id,
 	}
-
-	q := `SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level FROM groups WHERE id = $1`
+	q := `SELECT id, name, owner_id, parent_id, description, metadata, path, nlevel(path) as level, created_at, updated_at FROM groups WHERE id = $1`
 	if err := gr.db.QueryRowxContext(ctx, q, id).StructScan(&dbu); err != nil {
 		if err == sql.ErrNoRows {
-			return groups.Group{}, errors.Wrap(groups.ErrNotFound, err)
+			return auth.Group{}, errors.Wrap(auth.ErrGroupNotFound, err)
 
 		}
-		return groups.Group{}, errors.Wrap(errRetrieveDB, err)
+		return auth.Group{}, errors.Wrap(errRetrieveGroup, err)
 	}
-
-	return toGroup(dbu)
+	return gr.toGroup(dbu)
 }
 
-func (gr groupRepository) RetrieveAll(ctx context.Context, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
-	_, mq, err := getGroupsMetadataQuery("groups", gm)
+func (gr groupRepository) RetrieveAll(ctx context.Context, pm auth.PageMetadata) (auth.GroupPage, error) {
+	_, mq, err := getGroupsMetadataQuery("groups", pm.Metadata)
 	if err != nil {
-		return groups.GroupPage{}, errors.Wrap(errRetrieveDB, err)
+		return auth.GroupPage{}, errors.Wrap(errRetrieve, err)
 	}
+	cq := "SELECT COUNT(*) FROM groups"
 	if mq != "" {
+		cq = fmt.Sprintf("%s WHERE %s", cq, mq)
 		mq = fmt.Sprintf("AND %s", mq)
 	}
 
 	q := fmt.Sprintf(`SELECT id, owner_id, parent_id, name, description, metadata, path, nlevel(path) as level, created_at, updated_at FROM groups 
 					  WHERE nlevel(path) <= :level %s ORDER BY path`, mq)
-	cq := fmt.Sprintf("SELECT COUNT(*) FROM groups WHERE nlevel(path) <= :level %s", mq)
 
-	dbPage, err := toDBGroupPage("", "", "", "", level, gm)
+	dbPage, err := toDBGroupPage("", "", "", "", pm)
 	if err != nil {
-		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+		return auth.GroupPage{}, errors.Wrap(errRetrieveAll, err)
 	}
 
 	rows, err := gr.db.NamedQueryContext(ctx, q, dbPage)
 	if err != nil {
-		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+		return auth.GroupPage{}, errors.Wrap(errRetrieveAll, err)
 	}
 	defer rows.Close()
 
-	items, err := processRows(rows)
+	items, err := gr.processRows(rows)
 	if err != nil {
-		return groups.GroupPage{}, err
+		return auth.GroupPage{}, err
 	}
 
 	total, err := total(ctx, gr.db, cq, dbPage)
 	if err != nil {
-		return groups.GroupPage{}, errors.Wrap(errSelectDb, err)
+		return auth.GroupPage{}, errors.Wrap(errGetTotal, err)
 	}
 
-	page := groups.GroupPage{
+	page := auth.GroupPage{
 		Groups: items,
-		PageMetadata: groups.PageMetadata{
+		PageMetadata: auth.PageMetadata{
 			Total: total,
+			Size:  uint64(len(items)),
 		},
 	}
 
 	return page, nil
 }
 
-func (gr groupRepository) RetrieveAllParents(ctx context.Context, groupID string, level uint64, gm groups.Metadata) (groups.GroupPage, error) {
+func (gr groupRepository) RetrieveAllParents(ctx context.Context, groupID string, pm auth.PageMetadata) (auth.GroupPage, error) {
```

Comment:
Looks the same as RetriveAllChildren, except the query? Would it make sense to have private fn like retriveRelationship which will accept the query? Keep both RetriveAllParents, RetriveAllChildren on API level as adapters which will build and pass the query to retriveRelationship.